src/Storages/RemoteReadBufferCache.h

```diff
@@ -0,0 +1,219 @@
+#pragma once
+#include <mutex>
+#include <list>
+#include <set>
+#include <map>
+#include <memory>
+#include <filesystem>
+#include <Core/BackgroundSchedulePool.h>
+#include <Poco/Logger.h>
+#include <Common/LRUCache.h>
+#include <Common/ThreadPool.h>
+#include <IO/ReadBuffer.h>
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/createReadBufferFromFileBase.h>
+#include <IO/WriteBufferFromFile.h>
+#include <IO/WriteBufferFromFileBase.h>
+#include <IO/ReadBufferFromFileBase.h>
+#include <IO/ReadSettings.h>
+#include <IO/SeekableReadBuffer.h>
+#include <Storages/IRemoteFileMetadata.h>
+#include <condition_variable>
+#include <Interpreters/Context.h>
+#include <boost/core/noncopyable.hpp>
+
+
+namespace DB
+{
+enum class RemoteReadBufferCacheError : int8_t
+{
+    OK,
+    NOT_INIT = 10,
+    DISK_FULL = 11,
+    FILE_INVALID = 12,
+    END_OF_FILE = 20,
+};
+
+class RemoteCacheController
+{
+public:
+    enum LocalFileStatus
+    {
+        TO_DOWNLOAD = 0,
+        DOWNLOADING = 1,
+        DOWNLOADED  = 2,
+    };
+
+    RemoteCacheController(
+        IRemoteFileMetadataPtr file_metadata_,
+        const std::filesystem::path & local_path_,
+        size_t cache_bytes_before_flush_);
+    ~RemoteCacheController();
+
+    // recover from local disk
+    static std::shared_ptr<RemoteCacheController>
+    recover(const std::filesystem::path & local_path);
+
+    /**
+     * Called by LocalCachedFileReader, must be used in pair
+     * The second value of the return tuple is the local_path to store file.
+     */
+    std::unique_ptr<ReadBufferFromFileBase> allocFile();
+    void deallocFile(std::unique_ptr<ReadBufferFromFileBase> buffer);
+
+    /**
+     * when allocFile be called, count++. deallocFile be called, count--.
+     * the local file could be deleted only count==0
+     */
+    inline bool closable()
+    {
+        std::lock_guard lock{mutex};
+        //return opened_file_streams.empty() && remote_read_buffer == nullptr;
+        return opened_file_buffer_refs.empty() && file_status == DOWNLOADED;
+    }
+    void close();
+
+    /**
+     * called in LocalCachedFileReader read(), the reading process would be blocked until
+     * enough data be downloaded.
+     * If the file has finished download, the process would unblocked
+     */
+    RemoteReadBufferCacheError waitMoreData(size_t start_offset_, size_t end_offset_);
+
+    inline size_t size() const { return current_offset; }
+
+    inline const std::filesystem::path & getLocalPath() { return local_path; }
+    inline String getRemotePath() const { return file_metadata_ptr->getRemotePath(); }
+
+    inline UInt64 getLastModificationTimestamp() const { return file_metadata_ptr->getLastModificationTimestamp(); }
+    bool checkFileChanged(IRemoteFileMetadataPtr file_metadata_);
```

Comment:
```suggestion
    bool isModified(IRemoteFileMetadataPtr file_metadata_);
```
May be like this? what do you think?

Reply:
done